federationsender/queue/destinationqueue.go

```diff
@@ -33,92 +34,196 @@ import (
-	sentCounter        int
-	lastTransactionIDs []gomatrixserverlib.TransactionID
-	pendingEvents      []*gomatrixserverlib.HeaderedEvent
-	pendingEDUs        []*gomatrixserverlib.EDU
-	pendingInvites     []*gomatrixserverlib.InviteV2Request
+	rsProducer         *producers.RoomserverProducer        // roomserver producer
+	client             *gomatrixserverlib.FederationClient  // federation client
+	origin             gomatrixserverlib.ServerName         // origin of requests
+	destination        gomatrixserverlib.ServerName         // destination of requests
+	running            atomic.Bool                          // is the queue worker running?
+	wakeup             chan bool                            // wakes up a sleeping worker
+	statistics         *types.ServerStatistics              // statistics about this remote server
+	idleCounter        atomic.Uint32                        // how many ticks have we done nothing?
+	runningMutex       sync.RWMutex                         // protects the below
+	lastTransactionIDs []gomatrixserverlib.TransactionID    // protected by runningMutex
+	pendingPDUs        []*gomatrixserverlib.HeaderedEvent   // protected by runningMutex
+	pendingEDUs        []*gomatrixserverlib.EDU             // protected by runningMutex
+	pendingInvites     []*gomatrixserverlib.InviteV2Request // protected by runningMutex
+}
+
+// Start the destination queue if it needs to be started, or
+// otherwise signal to it that it should wake up from sleep.
+func (oq *destinationQueue) wake() {
+	if !oq.running.Load() {
+		go oq.backgroundSend()
+	} else {
+		select {
+		case oq.wakeup <- true:
+		default:
+		}
+	}
 }
 
 // Send event adds the event to the pending queue for the destination.
 // If the queue is empty then it starts a background goroutine to
 // start sending events to that destination.
 func (oq *destinationQueue) sendEvent(ev *gomatrixserverlib.HeaderedEvent) {
-	oq.runningMutex.Lock()
-	defer oq.runningMutex.Unlock()
-	oq.pendingEvents = append(oq.pendingEvents, ev)
-	if !oq.running.Load() {
-		go oq.backgroundSend()
+	if oq.statistics.Blacklisted() {
+		// If the destination is blacklisted then drop the event.
+		return
 	}
+	oq.runningMutex.Lock()
+	oq.pendingPDUs = append(oq.pendingPDUs, ev)
+	oq.runningMutex.Unlock()
+	oq.wake()
 }
 
 // sendEDU adds the EDU event to the pending queue for the destination.
 // If the queue is empty then it starts a background goroutine to
 // start sending events to that destination.
-func (oq *destinationQueue) sendEDU(e *gomatrixserverlib.EDU) {
-	oq.runningMutex.Lock()
-	defer oq.runningMutex.Unlock()
-	oq.pendingEDUs = append(oq.pendingEDUs, e)
-	if !oq.running.Load() {
-		go oq.backgroundSend()
+func (oq *destinationQueue) sendEDU(ev *gomatrixserverlib.EDU) {
+	if oq.statistics.Blacklisted() {
+		// If the destination is blacklisted then drop the event.
+		return
 	}
+	oq.runningMutex.Lock()
+	oq.pendingEDUs = append(oq.pendingEDUs, ev)
+	oq.runningMutex.Unlock()
```

Comment:
I was really hoping that we would replace this logic with a `chan` since after all this is exactly what we're emulating here, complete with an unbounded buffer.

Reply:
The problem with `chan` is that a buffered chan will occupy the memory of the full capacity, and we remove something from the `chan` by the nature of consuming it. We don't move things off the `pendingPDUs`, `pendingEDUs` until we have confirmed to have sent them, otherwise they remain there until the next retry.